WordPress/src/main/java/org/wordpress/android/ui/uploads/AutoSavePostIfNotDraftUseCase.kt

```diff
@@ -0,0 +1,108 @@
+package org.wordpress.android.ui.uploads
+
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode.BACKGROUND
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.PostActionBuilder
+import org.wordpress.android.fluxc.model.CauseOfOnPostChanged.RemoteAutoSavePost
+import org.wordpress.android.fluxc.model.LocalOrRemoteId.LocalId
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.store.PostStore
+import org.wordpress.android.fluxc.store.PostStore.OnPostChanged
+import org.wordpress.android.fluxc.store.PostStore.OnPostStatusFetched
+import org.wordpress.android.fluxc.store.PostStore.RemotePostPayload
+import org.wordpress.android.ui.uploads.AutoSavePostIfNotDraftResult.PostAutoSaved
+import org.wordpress.android.ui.uploads.AutoSavePostIfNotDraftResult.PostIsDraftInRemote
+import javax.inject.Inject
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+
+// TODO: Not be so verbose in naming
+
+private const val DRAFT_POST_STATUS = "draft"
+
+interface OnAutoSavePostIfNotDraftCallback {
+    fun handleAutoSavePostIfNotDraftResult(result: AutoSavePostIfNotDraftResult)
+}
+
+sealed class AutoSavePostIfNotDraftResult {
+    data class PostIsDraftInRemote(val post: PostModel) : AutoSavePostIfNotDraftResult()
+    data class PostAutoSaved(val post: PostModel) : AutoSavePostIfNotDraftResult()
+    // TODO: Add error result types
+}
+
+// TODO: Add documentation
+// TODO: Add unit tests
+class AutoSavePostIfNotDraftUseCase @Inject constructor(
+    private val dispatcher: Dispatcher,
+    private val postStore: PostStore
+) {
+    private val postStatusContinuations = HashMap<LocalId, Continuation<String>>()
+    private val autoSaveContinuations = HashMap<LocalId, Continuation<PostModel>>()
+
+    init {
+        dispatcher.register(this)
+    }
+
+    fun autoSavePostOrUpdateDraft(
+        remotePostPayload: RemotePostPayload,
+        callback: OnAutoSavePostIfNotDraftCallback
+    ) {
+        val localPostId = LocalId(remotePostPayload.post.id)
+        if (postStatusContinuations.containsKey(localPostId) ||
+                autoSaveContinuations.containsKey(localPostId)) {
+            // TODO: Consider canceling the previous job instead
+            // We are already handling this post
+            return
+        }
+        // TODO: What scope should we use for this?
+        GlobalScope.launch {
```

Comment:
Yeah, I thought so too, but maybe we can inject a scope to the use case directly? Even if what ends up being injected is the `GlobalScope` I think it's worth it.

I looked at it a bit, but since it's not a priority just left a todo to explore later.

Reply:
👍 I like the idea! That would definitely make it easier to reuse and less error prone.